docker/archive/multi-reader.go

```diff
@@ -0,0 +1,77 @@
+package archive
+
+import (
+	"context"
+
+	"github.com/containers/image/v5/docker/tarfile"
+	"github.com/containers/image/v5/types"
+)
+
+// MultiImageArchiveReader allows for reading a docker-archive with multiple
+// images.  It can be used in subsequent copy.Image operations when set as the
+// copy.Options.ImageSource.  Note that you need to call Next() before each
+// copy.Image operation.
+//
+// For using a MultiImageArchiveReader in copy.Image, set it as the
+// copy.Option.ImageSource.
+type MultiImageArchiveReader struct {
+	*archiveImageSource
+
+	manifests     []tarfile.ManifestItem
+	manifestIndex int
+}
+
+// NewMultiImageArchiveReader returns a new MultiImageArchiveReader based on the reference.
+func NewMultiImageArchiveReader(ctx context.Context, sys *types.SystemContext, ref types.ImageReference) (*MultiImageArchiveReader, error) {
+	path, destinationRef, err := parsePathAndNamedRef(ref.StringWithinTransport())
+	if err != nil {
+		return nil, err
+	}
+	archRef := archiveReference{
+		path:           path,
+		destinationRef: destinationRef,
+	}
+
+	src, err := newArchiveImageSource(ctx, sys, archRef)
+	if err != nil {
+		return nil, err
+	}
+
+	manifests, err := src.LoadTarManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	return &MultiImageArchiveReader{
+		archiveImageSource: src,
+		manifests:          manifests,
+		manifestIndex:      -1, // So we can call Next() before each copy.Image call.
+	}, nil
+}
+
+// Next selects the next image in the archive. False is returned when no image
+// is left.
+func (m *MultiImageArchiveReader) Next() bool {
```

Comment:
It’s a bit awkward that the caller has to manually track the same index value to obtain the tags from a separately-obtained copy of the manifest.